docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+* If the operator does not keep any state between windows, mark it with the **@Stateless** annotation. This results in efficiencies during checkpointing and recovery. The operator will not be checkpointed and always restored to the initial state
+* The checkpoint interval can be set using the **CHECKPOINT\_WINDOW\_COUNT** attribute which specifies the interval in terms of number of streaming windows.
+* If the correct functioning of the operator requires the **endWindow** method be called before checkpointing can happen, then the checkpoint interval should align with application window interval i.e., it should be a multiple of application window interval. In this case the operator should be marked with **OperatorAnnotation** and **checkpointableWithAppWindow** set to false. If the window intervals are configured by the user and they don’t align, it will result in a DAG validation error and application won’t launch.
+* In some cases the operator state related to a piece of data needs to be purged once that data is no longer required by the application, otherwise the state will continue to build up indefinitely. The platform provides a way to let the operator know about this using a callback listener called **CheckpointNotificationListener**. This listener has a callback method called **committed**, which is called by the platform from time to time with a window id that has been processed successfully by all the operators in the DAG and hence is no longer needed. The operator can delete all the state corresponding to window ids less than or equal to the provided window id.
+* Sometimes operators need to perform some tasks just before checkpointing. For example, filesystem operators may want to flush the files just before checkpoint so they can be sure that all pending data is written to disk and no data is lost if there is an operator failure just after the checkpoint and the operator restarts from the checkpoint. To do this the operator would implement the same *CheckpointNotificationListener* interface and implement the **beforeCheckpoint** method where it can do these tasks.
+* If the operator state is going to be large, checkpointing the entire state each time becomes unviable. Furthermore, the amount of state to keep could be larger than the amount of physical memory available. In those cases operator developers should use more efficient ways of maintaining state such as using **ManagedState** utility provided by the platform which uses a combination of in memory and disk cache to efficiently store and retrieve data in a performant and fault tolerant way. There are operators in the platform that use *ManagedState* and can be used as a reference on how to use this utility such as Dedup or Join operators.
+
+## Input Operators
+
+Input operators have additional requirements
+
+* The **emitTuples** method is called by platform giving it an opportunity to emit some data. There are some guidelines on how to implement this method that are important
```

Comment:
There are .... important => There are some important guidelines on how to implement this method: